vault/token_store.go

```diff
@@ -1148,45 +1227,93 @@ func (ts *TokenStore) revokeSalted(ctx context.Context, saltedID string, skipOrp
 
 	// Clear the secondary index if any
 	if entry.Parent != "" {
-		parentSaltedID, err := ts.SaltID(ctx, entry.Parent)
-		if err != nil {
-			return err
+		var parentIndex string
+		switch {
+		case entry.Version < 2:
+			// This is only here for backwards compatibility
+			parentSaltedID, err := ts.SaltID(ctx, entry.Parent)
+			if err != nil {
+				return err
+			}
+			parentIndex = parentPrefix + parentSaltedID + "/" + obfuscatedID
+		default:
+			parentIDHMAC, err := ts.hmac(ctx, entry.Parent)
+			if err != nil {
+				return err
+			}
+			parentIndex = parentPrefix + parentIDHMAC + "/" + obfuscatedID
 		}
-
-		path := parentPrefix + parentSaltedID + "/" + saltedID
-		if err = ts.view.Delete(ctx, path); err != nil {
+		if err = ts.view.Delete(ctx, parentIndex); err != nil {
 			return errwrap.Wrapf("failed to delete entry: {{err}}", err)
 		}
 	}
 
 	// Clear the accessor index if any
 	if entry.Accessor != "" {
-		accessorSaltedID, err := ts.SaltID(ctx, entry.Accessor)
-		if err != nil {
-			return err
+		var accessorIndex string
+		switch {
+		// This is only here for backwards compatibility
+		case entry.Version < 2:
+			accessorSaltedID, err := ts.SaltID(ctx, entry.Accessor)
+			if err != nil {
+				return err
+			}
+			accessorIndex = accessorPrefix + accessorSaltedID
+		default:
+			accessorHMAC, err := ts.hmac(ctx, entry.Accessor)
+			if err != nil {
+				return err
+			}
+			accessorIndex = accessorPrefix + accessorHMAC
 		}
-
-		path := accessorPrefix + accessorSaltedID
-		if err = ts.view.Delete(ctx, path); err != nil {
+		if err = ts.view.Delete(ctx, accessorIndex); err != nil {
 			return errwrap.Wrapf("failed to delete entry: {{err}}", err)
 		}
 	}
 
 	if !skipOrphan {
-		// Mark all children token as orphan by removing
-		// their parent index, and clear the parent entry.
+		// Mark all child tokens as orphans by removing their parent index, and
+		// clear the parent entry.
 		//
-		// Marking the token as orphan should be skipped if it's called by
-		// revokeTreeSalted to avoid unnecessary view.List operations. Since
-		// the deletion occurs in a DFS fashion we don't need to perform a delete
-		// on child prefixes as there will be none (as saltedID entry is a leaf node).
-		parentPath := parentPrefix + saltedID + "/"
-		children, err := ts.view.List(ctx, parentPath)
+		// Marking the token as orphan is the correct behavior in here since
+		// revokeTreeSalted will ensure that they are deleted anyways if it's not
+		// an explicit call to orphan the child tokens (the delete occurs at the
+		// leaf node and uses parent prefix, not entry.Parent, to build the tree
+		// for traversal).
+		idHMAC, err := ts.hmac(ctx, entry.ID)
```

Comment:
Don't we already have the `obfuscatedID` available to us? Why do we need to look it up again?

Reply:
This is a good catch. This avoided a lot of unnecessary code. Thanks! Please see a6b07133a81f6dcad2d4789a686649825abd0f51.